File: backend/app/services/operator_profile_service.py
from __future__ import annotations

import uuid
from datetime import datetime, timezone

from fastapi import HTTPException, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.operator import Operator, OperatorDocument


async def get_operator(db: AsyncSession, operator_id: str) -> Operator:
    result = await db.execute(select(Operator).where(Operator.id == operator_id))
    op = result.scalar_one_or_none()
    if not op:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Operator not found")
    return op


async def get_profile(db: AsyncSession, operator_id: str) -> Operator:
    return await get_operator(db, operator_id)


async def update_profile(db: AsyncSession, operator_id: str, changes: dict) -> Operator:
    op = await get_operator(db, operator_id)
    allowed = {"trade_name", "contact_email", "contact_phone", "hq_city"}
    # Legal-name / registration changes after approval require re-review
    locked_fields = {"name", "company_registration_no"}
    locked_being_changed = locked_fields.intersection(changes.keys())
    if locked_being_changed and op.onboarding_status == "approved":
        changes["onboarding_status"] = "re_review"
    for k, v in changes.items():
        if k in allowed | locked_fields | {"onboarding_status"}:
            setattr(op, k, v)
    await db.commit()
    await db.refresh(op)
    return op


async def submit_onboarding(db: AsyncSession, operator_id: str, data: dict) -> Operator:
    op = await get_operator(db, operator_id)
    if op.onboarding_status not in ("draft", "rejected", None):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Cannot submit from status '{op.onboarding_status}'",
        )
    for k, v in data.items():
        if hasattr(op, k):
            setattr(op, k, v)
    op.onboarding_status = "submitted"
    await db.commit()
    await db.refresh(op)
    return op


async def get_onboarding_status(db: AsyncSession, operator_id: str) -> dict:
    op = await get_operator(db, operator_id)
    return {
        "status": op.status,
        "onboarding_status": op.onboarding_status or "draft",
        "rejection_reason": op.rejection_reason,
    }


async def upload_certification(db: AsyncSession, operator_id: str, doc_type: str, file_url: str, expires_at) -> OperatorDocument:
    doc = OperatorDocument(
        id=str(uuid.uuid4()),
        operator_id=operator_id,
        doc_type=doc_type,
        file_url=file_url,
        expires_at=expires_at,
        status="pending",
        created_at=datetime.now(timezone.utc),
    )
    db.add(doc)
    await db.commit()
    await db.refresh(doc)
    return doc


async def upload_insurance(db: AsyncSession, operator_id: str, file_url: str, expires_at) -> OperatorDocument:
    doc = OperatorDocument(
        id=str(uuid.uuid4()),
        operator_id=operator_id,
        doc_type="insurance",
        file_url=file_url,
        expires_at=expires_at,
        status="pending",
        created_at=datetime.now(timezone.utc),
    )
    db.add(doc)
    await db.commit()
    await db.refresh(doc)
    return doc


async def update_payout_details(db: AsyncSession, operator_id: str, payout_account_ref: str) -> Operator:
    op = await get_operator(db, operator_id)
    op.payout_account_ref = payout_account_ref
    await db.commit()
    await db.refresh(op)
    return op


async def list_documents(db: AsyncSession, operator_id: str) -> list[OperatorDocument]:
    result = await db.execute(
        select(OperatorDocument)
        .where(OperatorDocument.operator_id == operator_id)
        .order_by(OperatorDocument.created_at.desc())
    )
    return list(result.scalars().all())

File: operator-panel/src/services/operatorProfileService.ts
import api from '../lib/axios'

export interface OperatorProfile {
  id: string
  name: string
  trade_name: string | null
  company_registration_no: string | null
  contact_email: string | null
  contact_phone: string | null
  hq_city: string | null
  cert_type: string | null
  status: string
  onboarding_status: string
  rejection_reason: string | null
  payout_account_ref: string | null
  insurance_expiry: string | null
  cert_expiry: string | null
}

export interface OnboardingSubmitRequest {
  name: string
  company_registration_no: string
  trade_name?: string
  contact_email: string
  contact_phone: string
  hq_city?: string
  cert_type?: string
}

export interface ProfileUpdate {
  trade_name?: string
  contact_email?: string
  contact_phone?: string
  hq_city?: string
}

export interface OnboardingStatus {
  status: string
  onboarding_status: string
  rejection_reason: string | null
}

export interface OperatorDocument {
  id: string
  doc_type: string
  file_url: string
  expires_at: string | null
  status: string
  review_notes: string | null
  created_at: string
}

export interface CertificationUpload {
  doc_type: string
  file_url: string
  expires_at?: string
}

export interface InsuranceUpload {
  file_url: string
  expires_at?: string
}

export const operatorProfileService = {
  getProfile: (): Promise<OperatorProfile> =>
    api.get<OperatorProfile>('/profile').then(r => r.data),

  updateProfile: (body: ProfileUpdate): Promise<OperatorProfile> =>
    api.patch<OperatorProfile>('/profile', body).then(r => r.data),

  submitOnboarding: (body: OnboardingSubmitRequest): Promise<OperatorProfile> =>
    api.post<OperatorProfile>('/onboarding/submit', body).then(r => r.data),

  getOnboardingStatus: (): Promise<OnboardingStatus> =>
    api.get<OnboardingStatus>('/onboarding/status').then(r => r.data),

  uploadCertification: (body: CertificationUpload): Promise<OperatorDocument> =>
    api.post<OperatorDocument>('/profile/certifications', body).then(r => r.data),

  uploadInsurance: (body: InsuranceUpload): Promise<OperatorDocument> =>
    api.post<OperatorDocument>('/profile/insurance', body).then(r => r.data),

  updatePayoutDetails: (payout_account_ref: string): Promise<OperatorProfile> =>
    api.patch<OperatorProfile>('/payout-details', { payout_account_ref }).then(r => r.data),

  listDocuments: (): Promise<OperatorDocument[]> =>
    api.get<OperatorDocument[]>('/profile/documents').then(r => r.data),
}
